File: 프로그래머스/3/17676. ［1차］ 추석 트래픽/［1차］ 추석 트래픽.py
def solution(lines):
    def check(start_time):
        end_time = start_time + 1000
        return sum(True for log in logs if log[0] < end_time and log[1] >= start_time)

    answer = 0
    logs = []
    
    for line in lines:
        d, t, s = line.split()
        t = t.split(':')
        s = float(s.replace('s', ''))
        end = (int(t[0]) * 3600 + int(t[1]) * 60 + float(t[2])) * 1000
        start = end - s * 1000 + 1
        logs.append((start, end))

    for target in logs:
        answer = max(answer, check(target[0]), check(target[1]))
    return answer

File: 프로그래머스/3/17676. ［1차］ 추석 트래픽/［1차］ 추석 트래픽.java
import java.math.BigDecimal;
import java.time.LocalTime;

class Solution {

    private static final long WINDOW_SIZE = 1_000L;

    public int solution(String[] lines) {
        int n = lines.length;
        long[] starts = new long[n];
        long[] ends = new long[n];

        for (int i = 0; i < n; i++) {
            String[] parts = lines[i].split(" ");

            long end = parseTime(parts[1]);
            long duration = parseDuration(parts[2]);

            ends[i] = end;
            starts[i] = end - duration + 1;
        }

        int answer = 0;

        for (long windowStart : ends) {
            long windowEnd = windowStart + WINDOW_SIZE - 1;
            int count = 0;

            for (int i = 0; i < n; i++) {
                if (starts[i] <= windowEnd && ends[i] >= windowStart) {
                    count++;
                }
            }

            answer = Math.max(answer, count);
        }

        return answer;
    }

    private long parseTime(String time) {
        return LocalTime.parse(time).toNanoOfDay() / 1_000_000;
    }

    private long parseDuration(String duration) {
        String seconds = duration.substring(0, duration.length() - 1);

        return new BigDecimal(seconds)
                .movePointRight(3)
                .longValueExact();
    }
}